### PR TITLE
Allow specification of parameter information on a per-vehicle-type basis

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
On Sat, 23 Sep 2017, Francisco Ferreira wrote:

>  *  add a new @Vehicles field, which when not present means it applies to
>     all vehicles and when it is present only applies to vehicles listed
>  *  in the vehicle params we check to see if a couple of required fields are
>     present and give an error if any is missing; we don't do the same in the
>     library parameters but I think we should and we can take advantage of
>     that: if a required parameter is missing but it existed for some other
>     vehicle, then it would mean it doesn't apply to the current vehicle
> 
> The first one has two advantages: explicitly says to which vehicles it
> applies and it is easier to code. The second one has the advantage that we
> don't need a new field. Personally, I prefer the first.

There's another option - 
https://github.com/ArduPilot/ardupilot/pull/6819/commits/ce1cd3955f43c71cd2528eb2940d272118b602ef#diff-dec3f64931c56a074db3501032fe7e6aR78

We coul dtry to understand that declaration, which would keep nicely keep 
the documentation aligned with the code.

I suggest we do this as a separate issue, however.

> >  prog_param_fields = re.compile(r"[ \t]*// @(\w+): (.*)")
> +# match e.g @Value[ArduCopter]: 0=Volcano, 1=Peppermint
> 
> It's {}, not [] (I actually prefer the later, but I just really want the
> comment to match the code 🙂)

Fixed.

> Instead of this here, can't we change opts.vehicle default to None and check
> if it is set (and doing it right after parsing arguments)? We can remove
> more code by doing that too.

Yes, of course.  I'd love to be able to spend time fixing the entire 
script, but I'm really just trying to fix this one problem in the script's 
output, not everything wrong with the script.  Pretty sure vast swathes of 
this code can go, but I'd need to understand this whole globbing thing 
*followed by this iteration over th results of that and it's not required 
for accomplishing the goal of this PR.

These changes would include making it library-like, single-vehicle, 
disentangling functions and code, and possibly reviving the 
recursive-param-metadata branch.

> +# prog_group_param = re.compile(r"@Param: (\w+).*((?:\n[ \t]*// @(\w+): (.*
> ))+)(?:\n\n|\n[ \t]+[A-Z])", re.MULTILINE)
> +prog_group_param = re.compile(r"@Param: (\w+).*((?:\n[ \t]*// @(\w+)(?:{([,
> \w]+)})?: (.*))+)(?:\n\n|\n[ \t]+[A-Z])", re.MULTILINE)
> 
> Any reason to leave the commented code?

Removed.

> Unrelated, but is there any reason why we have prog_param and
> prog_group_param? Shouldn't those two be equal? Even with these changes,
> making them equal means that we can add frame documentation differentiation
> in the future.

I don't think that would be a prereq, but yeah, they should be the same. 
I've added a patch for that.

> Can we separate vehicle name from the documentation? I'm pretty sure that
> someday Rover will change its folder name from APMRover2 to something else -
> although we have many scripts that will need to be adapted, having to change
> param documentation in code because the folder changed name would be much
> worse.

OK, I've added in a map for that.

I think we should have a final discussion on canonicalising everything and 
do the relevant renames.  Are we going to rename AntennaTracker to 
ArduTracker?  Are we going to change all of the directory names to just 
"Plane", "Sub", "Copter" etc?  Do we canonicalise on "Ardu%s"?  That sort 
of thing...

> My proposal is to have a dict that matches vehicle path to what we should
> have in documentation (and for that I propose: Copter, Plane, Rover, Sub,
> Tracker), so in code it would show as @Values{Copter}: ....

I've added that map in, JUST for this feature for the time being, and in a 
non-pretty manner.

> +                    error("param: unknown parameter metadata field %s" % fi
> eld[0])
> +#            print("matching %s" % field_text)
> 
> Left-over?

Made it a debug statement instead.

> +#                print("vehicle=%s field[1]=%s only_for_vehicles=%s\n" % (v
> ehicle.name,field[1], str(only_for_vehicles)))
> 
> Left-over?

Made it a debug statement instead.

> tagged param unknown parameter -> tagged param: unknown parameter
> 
> Just to make it consistent with the other messages.

Done

I think all of the param_parse.py patches could legitimately be squashed 
together.

**Comment by OXINARF** [comment]
> We coul dtry to understand that declaration, which would keep nicely keep
the documentation aligned with the code.
>
> I suggest we do this as a separate issue, however.

I won't be against going that way, but I'll point some things:

- maybe harder to code
- your argument about keeping code and documentation synced is good, but I predict the opposite will happen: new frame options are added to code and we forget to change the parser to support it or a new macro that also support frame options is added and the parser is forgotten; doing it with documentation will always remember people of the parser
- it doesn't cover Copter general/heli because we compile parameters in or out based on HELI_FRAME, not using the frame options (pretty sure we can change that, just pointing it out)

> I think we should have a final discussion on canonicalising everything and
do the relevant renames.  Are we going to rename AntennaTracker to
ArduTracker?  Are we going to change all of the directory names to just
"Plane", "Sub", "Copter" etc?  Do we canonicalise on "Ardu%s"?  That sort
of thing...

Yeah, I think that's part of a larger discussion on our naming. I'm pretty sure that will take weeks, if not months, to agree on, so I don't want this to be hold on that. At the same time we dissociate the documentation names from the folder names - I think the ones I proposed for documentation are quite consensual.

**Comment by OXINARF** [comment]
@jaxxzer Are you OK with this instead of your version?

File: Tools/autotest/param_metadata/param_parse.py [hunk 1]
```diff
@@ -23,14 +23,15 @@
 
 # Regular expressions for parsing the parameter metadata
 
-prog_param = re.compile(r"@Param: *(\w+).*((?:\n[ \t]*// @(\w+): (.*))+)(?:\n\n|\n[ \t]+[A-Z])", re.MULTILINE)
+prog_param = re.compile(r"@Param: (\w+).*((?:\n[ \t]*// @(\w+)(?:{([,\w]+)})?: (.*))+)(?:\n\n|\n[ \t]+[A-Z])", re.MULTILINE)
 
+# match e.g @Value: 0=Unity, 1=Koala, 17=Liability
 prog_param_fields = re.compile(r"[ \t]*// @(\w+): (.*)")
+# match e.g @Value{ArduCopter}: 0=Volcano, 1=Peppermint
```
Comment (jaxxzer) [inline Tools/autotest/param_metadata/param_parse.py]: Omg comment for the win thank you xD.

**Comment by jaxxzer** [comment]
It looks like this accomplishes the same thing as my PR, if that's the case I'm fine with this. Does this support specifying more than one @Vehicle{ArduCopter,ArduSub} ?

**Comment by peterbarker** [comment]
On Tue, 26 Sep 2017, jaxxzer wrote:

> I'm fine with this. Does this support specifying more than one
> @Vehicle{ArduCopter,ArduSub} ?

Yes, I swiped that bit :-)

**Comment by OXINARF** [comment]
@jaxxzer Like @peterbarker said, yes, we got that part from your PR, it was a great addition. We went with short names though: Copter, Plane, Rover, Sub and Tracker. So you write `@Values{Copter,Sub}`.

@peterbarker Am I correct that a space between vehicles (`@Values{Copter, Sub}`) wouldn't work? Should we add space to the regex and trim each token before comparing to the vehicle?

**Comment by OXINARF** [comment]
@peterbarker 
> I'm disinclined to have whitespace on the LHS of a "Foo: bar" thing - even if it is within {}

Agreed, but spaces **after** a separator is pretty common.

> OTOH, people *are* going to add it and get annoyed when their parameters
don't work out.  I've pushed up a patch which allows it.

It's worse than getting annoyed, we would just not catch these parameters (they wouldn't match) and we wouldn't know about it until someone found out they weren't in our wiki.

<br/>
@peterbarker @jaxxzer I think this is ready to merge (squash needed, but that's easy), but it would break two GCS: MissionPlanner (that does its own parsing) and APM Planner (that downloads the global *wrong* parameter file instead of vehicle ones). I'm thinking of opening issues for them but tagging @meee1 and @billbonney for now.

QGC doesn't automatically download our parameter files, but keeps versions of the XML file in their GitHub repo, so, although it won't benefit immediately, it won't be affected.

**Comment by jaxxzer** [comment]
Ok, I can patch/PR QGC when this is done.

**Comment by peterbarker** [comment]
@OXINARF How long are we going to hold this one up?  This is blocking the RCx_OPT change, which is blocking a couple of other changes.... :-)

How badly does this break MissionPlanner?  If we use the feature then the Values don't show?

I assume if we're not actually using it then it wont break.  Perhaps we could trim out the patches in this PR which which actually use the new feature so we can move on with getting the `RCx_OPT` stuff moving again?

**Comment by amilcarlucas** [comment]
Please prioritize this.

**Comment by OXINARF** [comment]
@peterbarker To be fair, I haven't made issues in MP and APM Planner repos yet.

> How badly does this break MissionPlanner? If we use the feature then the Values don't show?

MP codebase is quite hard to track, but I think that the vehicle specific tags won't be shown. Which means that you are correct that if we don't use it, then it shouldn't break MP.

> Perhaps we could trim out the patches in this PR which which actually use the new feature so we can move on with getting the RCx_OPT stuff moving again?

That doesn't work for APM Planner though.

**Comment by peterbarker** [comment]
@OXINARF This might work for APM Planner, 'though: https://github.com/ArduPilot/apm_planner/pull/1070

File: Tools/autotest/param_metadata/param_parse.py [hunk 2]
```diff
@@ -63,6 +67,10 @@ def error(str_to_print):
     vehicles.append(Vehicle(name, path))
     debug('Found vehicle type %s' % name)
 
+if len(vehicles) > 1:
+    print("Single vehicle only, please")
+    sys.exit(1)
```
[the remark this answers is not in the document]
Comment (amilcarlucas) [reply Tools/autotest/param_metadata/param_parse.py]: @peterbarker removing code is always good.

**Comment by peterbarker** [comment]
@amilcarlucas You still have a "changes requested" on this.  Are they going to block merging of this?

**Comment by amilcarlucas** [comment]
@OXINARF commented:
Instead of this here, can't we change opts.vehicle default to None and check if it is set (and doing it right after parsing arguments)? We can remove more code by doing that too.

I approve the PR, but I would like to see your reply to that one.

**Comment by peterbarker** [comment]
@amilcarlucas You're right, of course.

However, a natural consequence is to remove the loop iterating over the vehicle, which will adjust the indenting on the bulk of the script (or creation of a subroutine or class or...).  In any case major amounts of deltas. 

I do plan on cleaning this script up more based on it only ever handling a single vehicle - but that's not required for moving ahead with everything this changeset is blocking.

**Comment by khancyr** [comment]
ping ! it look like good to me

**Comment by amilcarlucas** [comment]
ping :) randy has approved it

**Comment by OXINARF** [comment]
@amilcarlucas Months ago... I need to re-look at this. Not sure if you are aware but all the pings generally aren't useful, it just makes me lose time looking at e-mails.

**Comment by gmorph** [comment]
Tested this morning with Peter B.  We actually uncovered and fixed an issue with MAVProxy regarding the display of the value parameters and once resolved this PR is working great.  Merging.

**Pull request merged**